clippy_lints/src/main_recursion.rs

```diff
@@ -0,0 +1,71 @@
+
+use syntax::symbol::sym;
+use rustc::lint::{LintArray, LintPass, LateLintPass, LateContext};
+use rustc::hir::{Crate, Expr, ExprKind, QPath};
+use rustc::hir::def::Res;
+use rustc::{declare_tool_lint, impl_lint_pass};
+
+use if_chain::if_chain;
+use crate::utils::{span_help_and_lint, is_entrypoint_fn};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for recursion using the entrypoint.
+    ///
+    /// **Why is this bad?** Apart from special setups (which we could detect following attributes like #![no_std]),
+    /// recursing into main() seems like an unintuitive antipattern we should be able to detect.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```no_run
+    /// fn main() {
+    ///     main();
+    /// }
+    /// ```
+    pub MAIN_RECURSION,
+    style,
+    "recursion using the entrypoint"
+}
+
+pub struct MainRecursion {
+    has_no_std_attr: bool
+}
+
+impl_lint_pass!(MainRecursion => [MAIN_RECURSION]);
+
+impl MainRecursion {
+    pub fn new() -> MainRecursion {
+        MainRecursion {
+            has_no_std_attr: false
+        }
+    }
+}
+
+impl LateLintPass<'_, '_> for MainRecursion {
+    fn check_crate(&mut self, _: &LateContext<'_, '_>, krate: &Crate) {
+        self.has_no_std_attr = krate.attrs.iter().any(|attr| attr.path == sym::no_std);
+    }
+
+    fn check_expr_post(&mut self, cx: &LateContext<'_, '_>, expr: &Expr) {
+        if self.has_no_std_attr {
+            return;
+        }
+
+        if_chain! {
+            if let ExprKind::Call(func, _) = &expr.node;
+            if let ExprKind::Path(path) = &func.node;
+            if let QPath::Resolved(_, ptr) = &path;
+            if let Res::Def(_, def_id) = &ptr.res;
+            if is_entrypoint_fn(cx, *def_id);
+            then {
+                span_help_and_lint(
+                    cx,
+                    MAIN_RECURSION,
+                    func.span,
+                    "recursing into `main()`",
```

Comment:
Could you change this, so that the name of the entrypoint function is displayed?

Pseudocode ahead!
```rust
&format!("recursing into entrypoint `{}`", func.name),
```

Reply:
`name` is not part of [rustc::hir::Expr](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/hir/struct.Expr.html)

Reply:
Yeah I know, that's why
> Pseudocode ahead!

😄 

Please find a way to get the name of the function. (Maybe that is possible with the `span` and `utils::snippet`)

Reply:
alright thanks